insonmnia/resource/pool.go

```
@@ -1,116 +1,187 @@
+	}
 }
 
-func (p *Pool) pollConsume(usage *Resources) error {
-	if usage.NumGPUs == -1 {
-		usage.NumGPUs = len(p.OS.GPU)
+func (p *pool) getFree() (*sonm.AskPlanResources, error) {
+	res := deepcopy.Copy(p.all).(*sonm.AskPlanResources)
+	usage, err := p.getUsage()
+	if err != nil {
+		return nil, err
 	}
+	err = res.Sub(usage)
+	if err != nil {
+		pool, _ := yaml.Marshal(res)
+		use, _ := yaml.Marshal(usage)
+		return &sonm.AskPlanResources{}, fmt.Errorf("resource pool inconsistency found - used resources are greater than available for scheduling(%s). pool - %s, used - %s",
+			err, pool, use)
+	}
+	return res, nil
+}
 
-	free := NewResources(
-		p.OS.LogicalCPUCount()-p.usage.NumCPUs,
-		int64(p.OS.RAM.Device.Available)-p.usage.Memory,
-		len(p.OS.GPU)-p.usage.NumGPUs,
-	)
+func (p *pool) getUsage() (*sonm.AskPlanResources, error) {
+	sum := &sonm.AskPlanResources{}
+	for _, askPlan := range p.used {
+		if err := sum.Add(askPlan); err != nil {
+			return nil, err
+		}
+	}
+	return sum, nil
+}
 
-	if usage.NumCPUs > free.NumCPUs {
-		return ErrNotEnoughCPU
+func (p *pool) pollConsume(resources *sonm.AskPlanResources) error {
+	available, err := p.getFree()
+	if err != nil {
+		return err
 	}
-	if usage.Memory > free.Memory {
-		return ErrNotEnoughMemory
+	err = available.Sub(resources)
+	if err != nil {
+		return fmt.Errorf("not enough resources - %s", err)
 	}
-	if usage.NumGPUs > free.NumGPUs {
-		return ErrNotEnoughGPU
+	return nil
+}
+
+func (p *pool) consume(ID string, resources *sonm.AskPlanResources) error {
+	if err := p.pollConsume(resources); err != nil {
+		return err
+	}
+	if _, ok := p.used[ID]; ok {
+		return fmt.Errorf("resources with ID %s has been already consumed", ID)
 	}
 
+	p.used[ID] = resources
 	return nil
 }
 
-func (p *Pool) Release(usage *Resources) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-
-	p.usage.NumCPUs -= usage.NumCPUs
-	p.usage.Memory -= usage.Memory
-	p.usage.NumGPUs -= usage.NumGPUs
+func (p *pool) release(ID string) error {
```

Just `delete(p.used, ID)`?

---

I think here it's a good idea to signal about pool inconsistency. If you don't care - you can ignore this error